cherab/tools/spectroscopy/polychromator.py

```diff
@@ -0,0 +1,163 @@
+
+# Copyright 2016-2021 Euratom
+# Copyright 2016-2021 United Kingdom Atomic Energy Authority
+# Copyright 2016-2021 Centro de Investigaciones Energéticas, Medioambientales y Tecnológicas
+#
+# Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the
+# European Commission - subsequent versions of the EUPL (the "Licence");
+# You may not use this work except in compliance with the Licence.
+# You may obtain a copy of the Licence at:
+#
+# https://joinup.ec.europa.eu/software/page/eupl5
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied.
+#
+# See the Licence for the specific language governing permissions and limitations
+# under the Licence.
+
+import numpy as np
+from raysect.optical import InterpolatedSF
+from raysect.optical.observer import RadiancePipeline0D
+
+from .instrument import SpectroscopicInstrument
+
+
+class PolychromatorFilter(InterpolatedSF):
+    """
+    Defines a symmetrical trapezoidal polychromator filter as a Raysect's InterpolatedSF.
+
+    :param float wavelength: Central wavelength of the filter in nm.
+    :param float window: Size of the filtering window in nm. Default is 3.
+    :param float flat_top: Size of the flat top part of the filter in nm.
+                           Default is None (equal to window).
+    :param str name: Filter name (e.g. "H-alpha filter"). Default is ''.
+    """
+
+    def __init__(self, wavelength, window=3., flat_top=None, name=''):
+
+        if wavelength <= 0:
+            raise ValueError("Argument 'wavelength' must be positive.")
+
+        if window <= 0:
+            raise ValueError("Argument 'window' must be positive.")
+
+        flat_top = flat_top or window - 1.e-15
+
+        if flat_top <= 0:
+            raise ValueError("Argument 'flat_top' must be positive.")
+        if flat_top > window:
+            raise ValueError("Argument 'flat_top' must be less or equal than 'window'.")
+        if flat_top == window:
+            flat_top = window - 1.e-15
```

Comment:
I agree that properties should return the exact values specified by the user. This has now been fixed. Also the code was wrong, the correct code is `flat_top = window * (1 - 1.e-15)`. However, the case `flat_top == window` is valid because it implements a special case of an ideal rectangular filter. The `wavelengths` arrays must not contain duplicate values to avoid ambiguity at the initialization of the `InterpolatedSF` caused by the sorting. Therefore, we need to decrease the flat top interval by a value not much greater than machine epsilon.